File: app/src/main/java/ht/pq/khanh/bus/RxBus.java
package ht.pq.khanh.bus;

import io.reactivex.Observable;
import io.reactivex.subjects.PublishSubject;

/**
 * Created by khanhpq on 9/25/17.
 */

public class RxBus {
    private static final RxBus INSTANCE = new RxBus();

    public static RxBus getInstance() {
        return INSTANCE;
    }

    private final PublishSubject<Object> bus = PublishSubject.create();

    public void send(final Object event) {
        bus.onNext(event);
    }

    public Observable<Object> toObservable() {
        return bus;
    }

    public <T> Observable<T> toObservable(Class<T> eventType)
    {
        return bus.ofType(eventType);
    }

    public boolean hasObservers() {
        return bus.hasObservers();
    }
}

File: app/src/main/java/ht/pq/khanh/bus/RxBus.kt
package ht.pq.khanh.bus

import io.reactivex.Observable
import io.reactivex.subjects.PublishSubject

/**
 * Created by khanhpq on 9/25/17.
 */

class RxBus private constructor(){

    private val bus = PublishSubject.create<Any>()

    fun send(event: Any) {
        bus.onNext(event)
    }

    fun toObservable(): Observable<Any> {
        return bus
    }

    fun <T> toObservable(eventType: Class<T>): Observable<T> {
        return bus.ofType(eventType)
    }

    fun hasObservers(): Boolean {
        return bus.hasObservers()
    }

    companion object {
        val instance = RxBus()
    }
}
